mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp

```diff
@@ -62,6 +62,43 @@ void ConcatOp::getCanonicalizationPatterns(RewritePatternSet &results,
   results.add<ConcatOptimization>(context);
 }
 
+struct SqrtReciprocalOptimization : public OpRewritePattern<tosa::PowOp> {
+  using OpRewritePattern<tosa::PowOp>::OpRewritePattern;
+  // Pattern that matches a Sqrt + Reciprocal to make it into a rsqrt.
+  // Sqrt is represented in tosa by a Pow so we check for Pow + reciprocal.
+  LogicalResult matchAndRewrite(tosa::PowOp op,
+                                PatternRewriter &rewriter) const override {
+    // Check that the PowOp has a single user
+    if (!op->hasOneUse())
+      return failure();
+
+    Operation* user = *op->user_begin();
+    // Check that this user is a reciprocal
+    if (!isa<tosa::ReciprocalOp>(user))
+      return failure();
+
+    // Check that the Pow op is an Sqrt - its second input should be the scale, 0.5 for Sqrt.
+    Operation* powScale = op.getInput2().getDefiningOp();
+    if (!powScale || !isa<tosa::ConstOp>(powScale))
+      return failure();
+
+    auto scale = cast<tosa::ConstOp>(powScale).getValue().cast<DenseElementsAttr>();
+    auto scaleValue = scale.getSplatValue<float>();
```

Comment:
The scale can also be non-splat, in that case we would trigger an assertion here. Can you add a check that the tensor is a splat and return failure otherwise?